File: frontend/apps/ios/Pantopus/Core/Design/Components/EnvelopeOcrBox.swift
//
//  EnvelopeOcrBox.swift
//  Pantopus
//
//  Bounding-box overlay drawn on top of the scanned envelope hero in the
//  A13.15 Disambiguate flow. Two tones — `clean` (solid sky frame, sky
//  tinted fill) and `unclear` (dashed amber frame, amber-tinted fill with
//  a soft water-stain radial). Caller supplies the bounding rect in
//  envelope-local coordinates; the primitive renders the overlay only,
//  so it composes inside any `.overlay(alignment: .topLeading) { … }`.
//

import SwiftUI

/// Tone of an `EnvelopeOcrBox` overlay.
///
/// - `clean` — solid 2pt sky border + sky-tinted fill. Used when the OCR
///   confidence is high enough to trust.
/// - `unclear` — dashed 2pt amber border + amber-tinted fill with a soft
///   radial water-stain texture. Used when the scan is smudged / damaged
///   so the OCR confidence is low.
///
/// Mirrors `EnvelopeOcrTone` on Android.
public enum EnvelopeOcrTone: Sendable, Equatable {
    case clean
    case unclear
}

/// Overlay box drawn on top of a scanned envelope preview. The optional
/// `label` is the tiny mono-pill that tags the box (e.g. `NAME · 97%`),
/// rendered just above the top-left corner of the box.
@MainActor
public struct EnvelopeOcrBox: View {
    private let rect: CGRect
    private let tone: EnvelopeOcrTone
    private let label: String?

    public init(rect: CGRect, tone: EnvelopeOcrTone, label: String? = nil) {
        self.rect = rect
        self.tone = tone
        self.label = label
    }

    public var body: some View {
        ZStack(alignment: .topLeading) {
            shape
            if let label {
                labelPill(label)
                    .offset(x: -1, y: -16)
            }
        }
        .frame(width: rect.width, height: rect.height, alignment: .topLeading)
        .offset(x: rect.minX, y: rect.minY)
        .allowsHitTesting(false)
        .accessibilityElement()
        .accessibilityLabel(accessibilityLabel)
    }

    private var shape: some View {
        ZStack {
            RoundedRectangle(cornerRadius: Radii.xs, style: .continuous)
                .fill(fillColor)
            if tone == .unclear {
                waterStain
            }
            RoundedRectangle(cornerRadius: Radii.xs, style: .continuous)
                .stroke(strokeColor, style: strokeStyle)
        }
    }

    private var waterStain: some View {
        RadialGradient(
            colors: [
                Theme.Color.warning.opacity(0.35),
                Theme.Color.warning.opacity(0.18),
                Color.clear
            ],
            center: .center,
            startRadius: 0,
            endRadius: max(rect.width, rect.height) * 0.6
        )
        .blendMode(.multiply)
        .clipShape(RoundedRectangle(cornerRadius: Radii.xs, style: .continuous))
    }

    private func labelPill(_ text: String) -> some View {
        Text(text)
            .font(.system(size: 8, weight: .bold, design: .monospaced))
            .tracking(0.3)
            .foregroundStyle(Theme.Color.appTextInverse)
            .padding(.horizontal, Spacing.s1)
            .padding(.vertical, 1)
            .background(
                RoundedRectangle(cornerRadius: 2, style: .continuous)
                    .fill(strokeColor)
            )
            .textCase(.uppercase)
    }

    private var fillColor: Color {
        switch tone {
        case .clean: Theme.Color.primary600.opacity(0.08)
        case .unclear: Theme.Color.warning.opacity(0.12)
        }
    }

    private var strokeColor: Color {
        switch tone {
        case .clean: Theme.Color.primary500
        case .unclear: Theme.Color.warning
        }
    }

    private var strokeStyle: StrokeStyle {
        switch tone {
        case .clean:
            StrokeStyle(lineWidth: 2, lineCap: .round)
        case .unclear:
            StrokeStyle(lineWidth: 2, lineCap: .round, dash: [4, 4])
        }
    }

    private var accessibilityLabel: String {
        let base = tone == .clean ? "OCR bounding box, clean" : "OCR bounding box, unclear"
        if let label { return "\(base): \(label)" }
        return base
    }
}

#Preview("OCR box on envelope") {
    let envelope = RoundedRectangle(cornerRadius: Radii.md, style: .continuous)
        .fill(Theme.Color.appSurfaceMuted)
        .frame(width: 320, height: 200)
    return VStack(spacing: Spacing.s5) {
        envelope.overlay(alignment: .topLeading) {
            EnvelopeOcrBox(
                rect: CGRect(x: 16, y: 60, width: 120, height: 22),
                tone: .clean,
                label: "name · 97%"
            )
        }
        envelope.overlay(alignment: .topLeading) {
            EnvelopeOcrBox(
                rect: CGRect(x: 16, y: 60, width: 120, height: 22),
                tone: .unclear,
                label: "name · 31%"
            )
        }
    }
    .padding(Spacing.s4)
    .background(Theme.Color.appBg)
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/components/EnvelopeOcrBox.kt
@file:Suppress("MagicNumber", "LongMethod", "MatchingDeclarationName", "ModifierMissing")

package app.pantopus.android.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.DpRect
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * Tone of an [EnvelopeOcrBox] overlay.
 *
 * - `Clean` — solid 2dp sky border + sky-tinted fill. Used when the OCR
 *   confidence is high enough to trust.
 * - `Unclear` — dashed 2dp amber border + amber-tinted fill with a soft
 *   radial water-stain texture. Used when the scan is smudged / damaged
 *   so the OCR confidence is low.
 */
enum class EnvelopeOcrTone { Clean, Unclear }

/**
 * Bounding-box overlay drawn on top of a scanned envelope preview. Caller
 * supplies the rect in envelope-local coordinates; the primitive renders
 * the overlay only, so it composes inside any positioned overlay.
 *
 * @param rect Bounding rect (offset + size in dp).
 * @param tone [EnvelopeOcrTone.Clean] or [EnvelopeOcrTone.Unclear].
 * @param label Optional uppercase mono pill rendered above the top-left
 *     corner (e.g. `"NAME · 97%"`).
 */
@Composable
fun EnvelopeOcrBox(
    rect: DpRect,
    tone: EnvelopeOcrTone,
    modifier: Modifier = Modifier,
    label: String? = null,
) {
    val accessibilityText =
        buildString {
            append(if (tone == EnvelopeOcrTone.Clean) "OCR bounding box, clean" else "OCR bounding box, unclear")
            if (label != null) append(": ").append(label)
        }
    Box(
        modifier =
            modifier
                .offset(x = rect.left, y = rect.top)
                .size(width = rect.right - rect.left, height = rect.bottom - rect.top)
                .semantics { contentDescription = accessibilityText },
    ) {
        BoxShape(tone = tone)
        if (label != null) {
            LabelPill(
                text = label,
                tone = tone,
                modifier = Modifier.offset(x = (-1).dp, y = (-16).dp),
            )
        }
    }
}

@Composable
private fun BoxShape(tone: EnvelopeOcrTone) {
    val strokeColor = strokeColor(tone)
    val fillColor = fillColor(tone)
    val shape = RoundedCornerShape(Radii.xs)
    val density = LocalDensity.current
    val strokePx = with(density) { 2.dp.toPx() }
    val dashOnPx = with(density) { 4.dp.toPx() }

    Box(
        modifier =
            Modifier
                .fillMaxSize()
                .clip(shape)
                .background(fillColor)
                .then(
                    if (tone == EnvelopeOcrTone.Unclear) {
                        Modifier.drawBehind {
                            drawWaterStain()
                        }
                    } else {
                        Modifier
                    },
                ).drawBehind {
                    val pathEffect =
                        if (tone == EnvelopeOcrTone.Unclear) {
                            PathEffect.dashPathEffect(floatArrayOf(dashOnPx, dashOnPx), 0f)
                        } else {
                            null
                        }
                    drawRoundRect(
                        color = strokeColor,
                        topLeft = Offset(strokePx / 2, strokePx / 2),
                        size =
                            Size(
                                width = size.width - strokePx,
                                height = size.height - strokePx,
                            ),
                        cornerRadius =
                            androidx.compose.ui.geometry.CornerRadius(
                                with(density) { Radii.xs.toPx() },
                            ),
                        style = Stroke(width = strokePx, pathEffect = pathEffect),
                    )
                },
    )
}

@Composable
private fun LabelPill(
    text: String,
    tone: EnvelopeOcrTone,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier =
            modifier
                .clip(RoundedCornerShape(2.dp))
                .background(strokeColor(tone))
                .padding(horizontal = Spacing.s1, vertical = 1.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = text.uppercase(),
            style =
                TextStyle(
                    fontFamily = FontFamily.Monospace,
                    fontWeight = FontWeight.Bold,
                    fontSize = 8.sp,
                    letterSpacing = 0.3.sp,
                ),
            color = PantopusColors.appTextInverse,
        )
    }
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawWaterStain() {
    val brush =
        Brush.radialGradient(
            colors =
                listOf(
                    PantopusColors.warning.copy(alpha = 0.35f),
                    PantopusColors.warning.copy(alpha = 0.18f),
                    Color.Transparent,
                ),
            center = Offset(size.width / 2f, size.height / 2f),
            radius = maxOf(size.width, size.height) * 0.6f,
        )
    drawRect(brush = brush, blendMode = BlendMode.Multiply)
}

private fun strokeColor(tone: EnvelopeOcrTone): Color =
    when (tone) {
        EnvelopeOcrTone.Clean -> PantopusColors.primary500
        EnvelopeOcrTone.Unclear -> PantopusColors.warning
    }

private fun fillColor(tone: EnvelopeOcrTone): Color =
    when (tone) {
        EnvelopeOcrTone.Clean -> PantopusColors.primary600.copy(alpha = 0.08f)
        EnvelopeOcrTone.Unclear -> PantopusColors.warning.copy(alpha = 0.12f)
    }
